integrations: Add Hubot, Zapier and IFTTT integrations to search

AasthaGupta opened the pull request:
Fixes: #7091

AasthaGupta commented:
This is an in-progress pull request.

AasthaGupta committed:
integrations: Add Hubot, Zapier and IFTTT integrations to search.

Fixes: #7091

AasthaGupta commented:
@hackerkid I have done it for Hubot integrations for now. I will do the same for Zapier and IFTTT. Is this okay?

rishig commented:
Thanks @AasthaGupta! The code seems reasonable, but is there a good way to test that this works? (In terms of just manual testing). Might make sense to add the list of Hubot integrations to this PR before we do a full review.

AasthaGupta commented:
@rishig The list of Hubot integrations were already there. In this PR, I included the code so that they appear only on search. I am planning to add the Zapier inegrations the same way Hubot integrations were there. What I am trying to say is that Zapier has ~1000 apps that it supports, is it the correct way to hard-code them into the code just like Hubot integrations have been?

rishig commented:
hm, maybe I'm not doing something correctly. Should "Instagram" be appering here?

![image](https://user-images.githubusercontent.com/890911/33959379-21524c70-dffc-11e7-9e76-09442bb45918.png)

AasthaGupta commented:
![screenshot from 2017-12-14 03-05-32](https://user-images.githubusercontent.com/15193262/33963827-687d2108-e07c-11e7-8350-b7f36f3fb1c6.png)
@rishig 
Instagram is not showing on search because of this issue [#7402](https://github.com/zulip/zulip/issues/7402). Instagram is a legacy hubot integration which is not included yet.
assembla is a hubot integration right now included.
Before any search, Assembla is not shown in the card.
![screenshot from 2017-12-14 03-05-32](https://user-images.githubusercontent.com/15193262/33963844-79c39604-e07c-11e7-8b23-bb188a23e5f9.png)

After searching for assembla, the card for it appears.
![screenshot from 2017-12-14 03-05-51](https://user-images.githubusercontent.com/15193262/33963818-5adf3784-e07c-11e7-98ff-82a3cc56de89.png)

rishig commented:
Perfect, thanks @AasthaGupta! Sending to Brock for review.

rishig commented:
Looking through Zapier's integrations, basically all of them seem useful. Maybe let's add them to a separate file, so they don't take over the main integrations file?

We'll also want separate wording for when the Zapier integration is either "Premium", "Beta", "Upcoming", or whatever other label they have, so we should keep that data when making the list.

AasthaGupta committed:
integrations: Add Zapier integrations.

AasthaGupta commented:
@rishig @timabbott @hackerkid 
After adding around 1000 integrations to the integrations list, the search is slowing down. It takes a lot of time to search throught the list. I think another approach should be taken to implement this.

rishig commented:
Good to know. We do search through ~1000 emoji in our emoji picker typeahead, and ~4K users in the user typeahead, so I think this should be possible. Maybe post in #frontend or ask Brock to see if anyone has ideas?

AasthaGupta commented:
![screenshot from 2017-12-19 21-17-58](https://user-images.githubusercontent.com/15193262/34165684-f59c9cb4-e502-11e7-8777-4042613c9a74.png)

AasthaGupta commented:
![screenshot from 2017-12-19 21-17-58](https://user-images.githubusercontent.com/15193262/34165684-f59c9cb4-e502-11e7-8777-4042613c9a74.png)

rishig commented:
Great! If there is both a Hubot and Zapier integration, let's just show the Zapier one.

AasthaGupta committed:
integrations: Add IFTTT integrations.

AasthaGupta commented:
![screenshot from 2017-12-20 00-50-36](https://user-images.githubusercontent.com/15193262/34174724-e99b1522-e51f-11e7-90c6-5cfbbc9e5268.png)

AasthaGupta commented:
@rishig What about the case when there is both IFTTT and Zapier integration?

showell commented:
@AasthaGupta The conflict here is due to a change I merged from you (HUBOT_INTEGRATIONS_LEGACY), so hopefully that will be an easy thing for you to fix.

Let's make the order of preference be Zapier, IFTTT, Hubot.  We can always change it later.

AasthaGupta committed:
integrations: Apps in the order of preference- Zapier, IFTTT, Hubot.

AasthaGupta commented:
@rishig @showell 
![screenshot from 2017-12-23 11-52-43](https://user-images.githubusercontent.com/15193262/34317594-4ceeefb2-e7d8-11e7-85d3-b18792d5af39.png)
![screenshot from 2017-12-23 11-54-19](https://user-images.githubusercontent.com/15193262/34317595-4e00dc3a-e7d8-11e7-8b23-54fe2df5441c.png)

rishig commented:
quick note, we can just say "gmail", rather than "gmail-zapier"

rishig commented on zerver/views/integrations.py:
```diff
@@ -19,14 +19,21 @@
+
+# This is bit of hack to give preference to integrations in
+# the order- Zapier, IFTTT, Hubot.
+# In the dictionary first Hubot is added then IFTTT and
```
you can remove the last three lines of this comment

rishig commented on zerver/views/integrations.py:
```diff
@@ -19,14 +19,21 @@
 for ifttt_integration in IFTTT_INTEGRATIONS:
     INTEGRATIONS[ifttt_integration.name] = ifttt_integration
 
+# add Zapier integrations into the dict
```
in general I wouldn't add comments like this into the code (i.e. comments that explain something that one can see by reading the code itself), since the comments can go out of date

rishig commented:
Thanks @AasthaGupta! A few notes:
* We'll have to fix the search slowness issue before this is mergable
* Currently if you search for something like "g..." you'll get mostly things that don't start with g. We should use the same algorithm we use for emoji search for determining the search order. 
* http://localhost:9991/integrations/doc/zapier seems to have gotten overwritten (i.e. the instructions for installing zapier).

AasthaGupta commented:
@rishig The search is linear for emoji search whereas search for apps is a fuzzy search so it is possible to that when you search for "g...", it shows things that don't start with g. I think it makes sense to use fuzzy search for integration because the user might search for an integration using a term that can be anywhere in the integration's name. Also the search is faster when using fuzzy search compared to linear search for integrations.

Search is slow only if the user's search query has less characters (~1-5) because then it will render lot of integration cards in the result which slows down the search. If there are more characters in the query, then search results are less and there is less rendering to do.

rishig commented:
Both emoji search and this are linear and fuzzy, but it sounds like rendering the cards is slowing down the search. Let's say something like:
* If any integrations match as a prefix (e.g. "goo" matches "google", but not "xygoogle" or "xy google") then just show the ones that match the prefix. 
* If no integrations match the prefix, then show all that match any substring (what we're calling fuzzy search)

Also linear search is a good concept to understand; feel free to ask about it in dev help!

AasthaGupta commented:
This way won't we have to iterate through the list of apps twice? Once to search by prefix and if the results are none then to check by substring.
Also I understand that both the searches are linear. I used the wrong term over there. What I meant was that it is doing substring search using `indexOf() >= 0`.

showell commented:
I'm not convinced that either search algorithm is the bottleneck here.  I think this may be a rendering thing.

String searches are generally super fast.

AasthaGupta commented:
yes. rendering is the bottleneck. maybe we can limit the number of search results shown?

AasthaGupta committed:
integrations: Modify search process of integrations.

If any integrations matches as a prefix then, show those
integrations. If no integration matches the prefix,then
show all those integrations that match as substring.

AasthaGupta commented:
@rishig 
I have updated the pull request with suggested changes. Now if the query matches as prefix, then those results are shown otherwise results of matches as substring are shown.

rishig commented:
Thanks @AasthaGupta! A few comments below
* One thing I noticed is that we have both "google hangouts" (hubot) and "google hangouts chat" (zapier). Possibly we should just remove the hubot one from the list?
* Another thing is that the doc pages seem to be broken for zapier integrations? e.g. http://localhost:9991/integrations/doc/google-hangouts-chat looks like
![image](https://user-images.githubusercontent.com/890911/35291154-47da8bce-0021-11e8-9632-2228c43553f1.png)
* We shouldn't use "Integration frameworks" as the category for Zapier integrations.
![image](https://user-images.githubusercontent.com/890911/35291184-6769f22c-0021-11e8-8602-0551e8f0f3e3.png)
* Clicking a category link is very slow, and gives odd results (e.g. a bunch of integration stuff that starts with A, and then everything else)

Possibly this just means we need to categorize the 750 zapier integrations?

AasthaGupta commented:
@rishig

-  Yes, I think we can remove 'Google Hangouts' from Hubot.

-  For now, a generic doc is presented for Zapier integrations. Does something need to be changed?

- Should I classify each of the 1000 Zapier integrations or maybe they can be classied as 'Zapier integrations'?

- I will look into this.

rishig commented:
* a generic doc is great, but the generic doc seems to be missing some lines at the end, and missing the logo? 
* yeah, I think we will want to classify the 1000 zapier integrations. That may also help with the performance problems. If you put them into a google sheet, we can maybe get some help from czo.

AasthaGupta commented:
@rishig 
I have put them all in this google sheet.
https://docs.google.com/spreadsheets/d/1EESlmHAL-FVHJKJl4-DUntlXK_ScGclofSRKEqoDlUM/edit?usp=sharing

rishig commented:
Great! How easy would it be to add the category information that zapier provides as well? E.g. 
https://zapier.com/apps/categories/marketing

showell commented:
@rishig I'm un-assigning myself here, as I think we want Tim to handle this, since it has performance implications and I haven't really been tracking it.

AasthaGupta commented:
@rishig 
I have added the category information that the zapier provides into the doc.
Also, It is possible to get the logo images from zapier.

hackerkid commented:
@timabbott Can you take a look at this? @AasthaGupta Can you rebase meanwhile?

AasthaGupta commented:
@zulipbot label "needs review"
I have already rebased it.
